### PR TITLE
[pkg/stanza] Move core time parsing capabilities to coreinternal

### PR DESCRIPTION
**Pull request by TylerHelmuth**
**Description:** 
Moves strptime and native Go format parsing from stanza's `TimeParser` struct to `coreinternal` so it can be used by OTTL.

To keep the PR as small as possible and to reduce risk only the strptime and go parsing capabilities are shared.  Maybe more can be moved in the future, but these are currently the only pieces OTTL needs.  The implementation in `timeutils` also tries to be as similar to the original implementation as possible.  

The `TimeParser` API is unchanged so this should not be a breaking change.

**Issue**
Unblocks https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/22811, which needs this functionality to parse time the same way as stanza.

### CODE REVIEW

**Review by djaglowski — APPROVED** [review]

File: internal/coreinternal/timeutils/parser.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+		// If "location" is specified, it must be in the local timezone database
+		loc, err := time.LoadLocation(*location)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load location %s: %w", *location, err)
+		}
+		return loc, nil
+	}
+
+	if layout != nil && strings.HasSuffix(*layout, "Z") {
+		// If a timestamp ends with 'Z', it should be interpretted at Zulu (UTC) time
```
Comment (crobert-1) [inline internal/coreinternal/timeutils/parser.go:29]:
```suggestion
		// If a timestamp ends with 'Z', it should be interpreted at Zulu (UTC) time
```

File: internal/coreinternal/timeutils/parser.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	strptime "github.com/observiq/ctimefmt"
+)
+
+func StrptimeToGo(layout string) (string, error) {
```
Comment (crobert-1) [inline internal/coreinternal/timeutils/parser.go:14]: What's the benefit of creating this function over just directly calling `strptime.ToNative`? Are we just trying to centralize all of the necessary time logic in this package? It just seems unnecessary.

File: internal/coreinternal/timeutils/parser.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+	return strptime.ToNative(layout)
+}
+
+func GetLocation(location *string, layout *string) (*time.Location, error) {
+	if location != nil && *location != "" {
+		// If "location" is specified, it must be in the local timezone database
```
Comment (crobert-1) [inline internal/coreinternal/timeutils/parser.go:20]:
```suggestion
		// If location is specified, it must be in the local timezone database
```
Just a thought, feel free to ignore.

File: pkg/stanza/operator/helper/time.go [hunk 4]
```diff
@@ -147,63 +147,26 @@ func (t *TimeParser) Parse(entry *entry.Entry) error {
...
-		entry.Timestamp = setTimestampYear(timeValue)
+		entry.Timestamp = timeutils.SetTimestampYear(timeValue)
 	case GotimeKey:
-		timeValue, err := t.parseGotime(value)
+		timeValue, err := timeutils.ParseGoTime(t.Layout, value, t.location)
 		if err != nil {
 			return err
 		}
-		entry.Timestamp = setTimestampYear(timeValue)
+		entry.Timestamp = timeValue
```
[the remark this answers is not in the document]
Comment (TylerHelmuth) [reply pkg/stanza/operator/helper/time.go:157]: Good call.  Someday when all cases are in timeutils it won't look as weird.

File: internal/coreinternal/timeutils/parser.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+// SetTimestampYear sets the year of a timestamp to the current year.
+// This is needed because year is missing from some time formats, such as rfc3164.
+func SetTimestampYear(t time.Time) time.Time {
+	if t.Year() > 0 {
+		return t
+	}
+	n := Now()
+	d := time.Date(n.Year(), t.Month(), t.Day(), t.Hour(), t.Minute(), t.Second(), t.Nanosecond(), t.Location())
+	// If the timestamp would be more than 7 days in the future using this year,
+	// assume it's from last year.
```
Comment (crobert-1) [inline internal/coreinternal/timeutils/parser.go:90]:
```suggestion
	// Assume the timestamp is from last year if its month and day are
	// more than 7 days past the current date.
	// i.e. If today is January 1, but the timestamp is February 1, it's safe
	// to assume the timestamp is from last year.
```
I know you're just copying this method, so feel free to ignore. It took me a while to understand this comment and the logic in the if statement here, so I'm just trying to come up with something a bit more clear.

File: internal/coreinternal/timeutils/parser.go (hunk 2 above)
Comment (TylerHelmuth) [reply internal/coreinternal/timeutils/parser.go]: Thats fair.  I was too focused on how stanza was doing things and not what will be most useful in a general sense.  Replacing this with a ParseStrptime function for OTTL or any other component to use makes sense.

**Review by crobert-1 — APPROVED** [review]

File: .chloggen/main.yaml
Comment (evan-bradley) [inline .chloggen/main.yaml:1]: Is this intentional? The PR has a `Skip Changelog` label and no OTTL code has changed in this PR.

Comment (TylerHelmuth) [reply .chloggen/main.yaml:1]: When I first created the PR I had some commits from https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/22811 to test the new shared code and all the labels and reviews got added by our CI.

Comment (TylerHelmuth) [reply .chloggen/main.yaml:1]: Even after your comment I missed the changelog lol it is now removed

**Review by evan-bradley — APPROVED** [review]

**Pull request merged**